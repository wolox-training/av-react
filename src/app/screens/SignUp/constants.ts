export const SIGN_UP_FIELDS = {
  firstName: 'firstName',
  lastName: 'lastName',
  email: 'email',
  password: 'password',
  passwordConfirmation: 'passwordConfirmation'
};
