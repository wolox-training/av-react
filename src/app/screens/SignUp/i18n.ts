import i18next from 'i18next';

i18next.addResources('es', 'SignUp', {
  logoAlt: 'WoloxLogo',
  firstName: 'Nombre',
  lastName: 'Apellido',
  email: 'Email',
  password: 'Password',
  signUp: 'Sign up',
  login: 'Login',
  confirmPassword: 'Confirmación de Password',
  required: 'Requerido',
  invalidEmail: 'Email inválido',
  passwordsNoMatch: 'Los passwords no coinciden'
});
