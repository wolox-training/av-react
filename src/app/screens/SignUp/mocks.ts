export const OK_SIGNUP_API_RESPONSE = {
  headers: {
    'access-token': 'test'
  },
  data: {
    email: 'test@test.com',
    firstName: 'test',
    id: 1,
    lastName: 'test',
    locale: 'es'
  },
  ok: true,
  originalError: null,
  problem: null,
  status: 201
};
