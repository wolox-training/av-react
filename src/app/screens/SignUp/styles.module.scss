@import 'variables/colors';

.signup-form {
  background-color: $wild-sand;
  border-top: 8px solid $cerulean;
  min-height: 634px;
  margin-top: 51px;
  text-align: left;
  width: 300px;

  .logo {
    height: 70px;
    margin: 28px 42.5px 0 45px;
    width: 212px;
  }

  .custom-label {
    align-self: flex-start;
    color: $black;
    font-size: 16px;
    line-height: 24px;
  }

  .custom-button {
    border-radius: 5px;
    font-size: 16px;
    height: 32px;

    &:disabled {
      background-color: $silver;
      border: 1px solid $dusty-gray;
      color: $dove-gray;
      cursor: not-allowed;
      pointer-events: none;
    }
  }

  .signup-button-container {
    margin: 24px 24px 0;

    &::after {
      content: '';
      background-color: $silver-chalice;
      display: block;
      height: 2px;
      margin-top: 16px;
    }
  }

  .signup-button {
    background-color: $earls-green;
    color: $white;
  }

  .login-button {
    background-color: $white;
    border: 1px solid $earls-green;
    color: $earls-green;
    margin: 16px 24px 24px;
  }

  .loading {
    margin-top: 16px;
  }
}
