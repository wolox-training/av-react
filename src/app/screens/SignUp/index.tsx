import React from 'react';
import i18next from 'i18next';
import { useForm } from 'react-hook-form';
import { Link } from 'react-router-dom';

import { User } from '~utils/types';
import { signUp } from '~services/UserService';
import CustomErrorDisplayer from '~components/CustomErrorDisplayer';
import Loading from '~components/Spinner/components/loading';
import CustomInput from '~components/CustomInput';
import { PATHS } from '~constants/paths';
// eslint-disable-next-line import/namespace
import { useLazyRequest } from '~hooks/useRequest';

import WoloxImg from '../Assets/wolox-logo.png';

import { SIGN_UP_FIELDS } from './constants';
import styles from './styles.module.scss';

interface FormData extends User {
  passwordConfirmation?: string;
}

export default function SignUp() {
  const { register, handleSubmit, watch, errors } = useForm<FormData>({ mode: 'all' });
  const [, loading, error, sendRequest] = useLazyRequest({
    request: signUp
  });
  const onSubmit = handleSubmit(data => {
    data.locale = i18next.language;
    sendRequest(data);
  });

  return (
    <div className="column center">
      <form onSubmit={onSubmit} className={`column ${styles.signupForm}`}>
        <img src={WoloxImg} alt={i18next.t('SignUp:logoAlt') as string} className={styles.logo} />
        <CustomInput
          name={SIGN_UP_FIELDS.firstName}
          inputRef={register({
            required: { value: true, message: i18next.t('SignUp:required') }
          })}
          error={errors.firstName?.message}
          label={i18next.t('SignUp:firstName') as string}
          labelClassName={styles.customLabel}
        />
        <CustomInput
          name={SIGN_UP_FIELDS.lastName}
          inputRef={register({
            required: { value: true, message: i18next.t('SignUp:required') }
          })}
          error={errors.lastName?.message}
          label={i18next.t('SignUp:lastName') as string}
          labelClassName={styles.customLabel}
        />
        <CustomInput
          name={SIGN_UP_FIELDS.email}
          inputRef={register({
            required: { value: true, message: i18next.t('SignUp:required') },
            pattern: {
              value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
              message: i18next.t('SignUp:invalidEmail')
            }
          })}
          error={errors.email?.message}
          type="email"
          label={i18next.t('SignUp:email') as string}
          labelClassName={styles.customLabel}
        />
        <CustomInput
          name={SIGN_UP_FIELDS.password}
          inputRef={register({
            required: { value: true, message: i18next.t('SignUp:required') }
          })}
          error={errors.password?.message}
          type="password"
          label={i18next.t('SignUp:password') as string}
          labelClassName={styles.customLabel}
        />
        <CustomInput
          name={SIGN_UP_FIELDS.passwordConfirmation}
          inputRef={register({
            required: { value: true, message: i18next.t('SignUp:required') },
            validate: value =>
              value === watch(SIGN_UP_FIELDS.password) || (i18next.t('SignUp:passwordsNoMatch') as string)
          })}
          error={errors.passwordConfirmation?.message}
          type="password"
          label={i18next.t('SignUp:confirmPassword') as string}
          labelClassName={styles.customLabel}
        />
        <div className={`column ${styles.signupButtonContainer}`}>
          <button
            className={`${styles.customButton} ${styles.signupButton} `}
            type="submit"
            disabled={loading}
          >
            {i18next.t('SignUp:signUp')}
          </button>
        </div>
        <Link to={PATHS.login} className="row">
          <button
            className={`${styles.customButton} ${styles.loginButton}  full-width`}
            type="button"
            disabled={loading}
          >
            {i18next.t('SignUp:login')}
          </button>
        </Link>
      </form>
      {loading && <Loading className={styles.loading} />}
      {error?.errorData?.errors.fullMessages.length && (
        <CustomErrorDisplayer errors={error.errorData?.errors.fullMessages} />
      )}
    </div>
  );
}
