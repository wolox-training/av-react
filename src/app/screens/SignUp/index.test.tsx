import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
<<<<<<< HEAD
import '@testing-library/jest-dom';
=======
>>>>>>> main

import SignUp from './';

const FORM_FIELDS_AMOUNT = 5;
const MAX_FORM_FIELDS_ERRORS = 5;
const MINIMUM_ALERT_LENGTH = 1;
const FORM_BUTTONS_AMOUNT = 2;
const PASSWORD = 'test';
const WRONG_CONFIRM_PASSWORD = 'wrongPassword';

describe('Signup test', () => {
  test('Should match snapshot', () => {
    const { container } = render(<SignUp />);
    expect(container).toMatchSnapshot();
  });

  test('Should render 5 form inputs', () => {
    const { container } = render(<SignUp />);
    expect(container.querySelectorAll('input').length).toBe(FORM_FIELDS_AMOUNT);
  });

  test('Should render 2 buttons', () => {
    render(<SignUp />);
    expect(screen.getAllByRole('button')).toHaveLength(FORM_BUTTONS_AMOUNT);
  });

  test('Wolox logo should be render and contain alt text', () => {
    render(<SignUp />);
    expect(screen.getByAltText('SignUp:logoAlt')).toBeInTheDocument();
  });

  test('Should render email error message', async () => {
    render(<SignUp />);
    userEvent.type(screen.getByLabelText('SignUp:email'), PASSWORD);
    await waitFor(() => expect(screen.getAllByRole('alert').length).toBe(MINIMUM_ALERT_LENGTH));
  });

  test('Should render required message for form fields', async () => {
    render(<SignUp />);
    fireEvent.click(screen.getByText('SignUp:signUp'));
    await waitFor(() => expect(screen.getAllByRole('alert').length).toBe(MAX_FORM_FIELDS_ERRORS));
  });

  test('Should render password match error', async () => {
    render(<SignUp />);
    userEvent.type(screen.getByLabelText('SignUp:password'), PASSWORD);
    userEvent.type(screen.getByLabelText('SignUp:confirmPassword'), WRONG_CONFIRM_PASSWORD);
    await waitFor(() => expect(screen.getAllByRole('alert').length).toBe(1));
  });
});
