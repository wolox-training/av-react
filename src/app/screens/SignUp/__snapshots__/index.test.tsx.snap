// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup test Should match snapshot 1`] = `
<div>
  <div
    class="column center"
  >
    <form
      class="column signupForm"
    >
      <img
        alt="SignUp:logoAlt"
        class="logo"
        src="wolox-logo.png"
      />
      <div
        class="inputContainer column"
      >
        <label
          class="customLabel"
          for="firstName"
        >
          SignUp:firstName
        </label>
        <input
          class="input full-width"
          id="firstName"
          name="firstName"
          type="text"
        />
      </div>
      <div
        class="inputContainer column"
      >
        <label
          class="customLabel"
          for="lastName"
        >
          SignUp:lastName
        </label>
        <input
          class="input full-width"
          id="lastName"
          name="lastName"
          type="text"
        />
      </div>
      <div
        class="inputContainer column"
      >
        <label
          class="customLabel"
          for="email"
        >
          SignUp:email
        </label>
        <input
          class="input full-width"
          id="email"
          name="email"
          type="email"
        />
      </div>
      <div
        class="inputContainer column"
      >
        <label
          class="customLabel"
          for="password"
        >
          SignUp:password
        </label>
        <input
          class="input full-width"
          id="password"
          name="password"
          type="password"
        />
      </div>
      <div
        class="inputContainer column"
      >
        <label
          class="customLabel"
          for="passwordConfirmation"
        >
          SignUp:confirmPassword
        </label>
        <input
          class="input full-width"
          id="passwordConfirmation"
          name="passwordConfirmation"
          type="password"
        />
      </div>
      <div
        class="column signupButtonContainer"
      >
        <button
          class="customButton signupButton "
          type="submit"
        >
          SignUp:signUp
        </button>
      </div>
      <a
        class="row"
        href="/"
      >
        <button
          class="customButton loginButton  full-width"
          type="button"
        >
          SignUp:login
        </button>
      </a>
    </form>
  </div>
</div>
`;
